### PR TITLE
Provide an arbitrary data persistence mechanism on storage instances

### PR DESCRIPTION
**Pull request by mikehaney24**
This could allow most singletons to no longer need NSCoding, they just store the individual fields they need with an appropriately unique key.

#no-changelog

### CODE REVIEW

File: GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m [hunk 1]
```diff
@@ -187,6 +208,39 @@ - (void)removeEvents:(NSSet<NSNumber *> *)eventIDs {
...
+  dispatch_async(_storageQueue, ^{
+    NSError *error;
+    NSString *dataPath = [[[self class] libraryDataPath] stringByAppendingPathComponent:key];
+    if (data && data.length > 0) {
+      [data writeToFile:dataPath options:NSDataWritingAtomic error:&error];
+      if (onComplete) {
+        onComplete(error);
+      }
+    } else {
+      if ([[NSFileManager defaultManager] fileExistsAtPath:dataPath]) {
```
Comment (paulb777) [inline GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m:234]: Would the API be easier to use if there was a `removeLibraryData` method instead of a zero lengthed `NSData`?

Either way, please add a test for this path.

File: GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m [hunk 2]
```diff
@@ -51,6 +51,27 @@ + (NSString *)archivePath {
   return archivePath;
 }
 
++ (NSString *)libraryDataPath {
```
Comment (maksymmalyhin) [inline GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m:54]: nit: Since the method has a side effect (may create a directory) it may be useful to add mention about it in the method name to improve readability, e.g. `libraryDataPathCreateIfNotExists`. I don't have a strong opinion here, so leave it up to you.

File: GoogleDataTransport/GDTCORLibrary/Public/GDTCORStorageProtocol.h [hunk 3]
```diff
@@ -36,6 +38,31 @@ NS_ASSUME_NONNULL_BEGIN
...
+- (void)storeLibraryData:(NSData *)data
+                  forKey:(NSString *)key
+              onComplete:(void (^)(NSError *_Nullable error))onComplete;
+
+/** Retrieves the stored data for the given key.
+ *
+ * @param key The key corresponding to the desired data.
+ * @param onComplete The callback to invoke with the data once it's retrieved.
+ */
+- (void)libraryDataForKey:(NSString *)key onComplete:(void (^)(NSData *_Nullable data))onComplete;
```
Comment (maksymmalyhin) [inline GoogleDataTransport/GDTCORLibrary/Public/GDTCORStorageProtocol.h:56]: Are there any potential read errors we may want to expose in the completion?

File: GoogleDataTransport/GDTCORLibrary/Public/GDTCORStorageProtocol.h [hunk 4]
```diff
@@ -36,6 +38,31 @@ NS_ASSUME_NONNULL_BEGIN
 /** Removes the events from storage. */
 - (void)removeEvents:(NSSet<NSNumber *> *)eventIDs;
 
+/** Persists the given data with the given key.
```
Comment (maksymmalyhin) [inline GoogleDataTransport/GDTCORLibrary/Public/GDTCORStorageProtocol.h:41]: The new methods look unrelated to the previously existing ones. Do you think a separate protocol (e.g. `GDTCORKeyValueStorage`) and implementation may be a better abstraction from this perspective?

File: GoogleDataTransport/GDTCORLibrary/Public/GDTCORStorageProtocol.h [hunk 5]
```diff
@@ -36,6 +38,31 @@ NS_ASSUME_NONNULL_BEGIN
...
+ *
+ * @param data The data to store.
+ * @param key The unique key to store it to.
+ * @param onComplete An block to be run when storage of the data is complete.
+ */
+- (void)storeLibraryData:(NSData *)data
```
Comment (maksymmalyhin) [inline GoogleDataTransport/GDTCORLibrary/Public/GDTCORStorageProtocol.h:47]: I wonder what is the intended usage of the storage. Is the client still expected to use `NSKeyedArchiver` to create data from e.g. an `NSArray` instance? If so, do you think we may use `id<NSSecureCoding>` instead of `NSData` type for values and encapsulate `NSKeyed(Un)Archiver` logic in the storage?

File: GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m (hunk 1 above)
Comment (mikehaney24) [reply GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m]: Done.

File: GoogleDataTransport/GDTCORLibrary/Public/GDTCORStorageProtocol.h (hunk 4 above)
Comment (mikehaney24) [reply GoogleDataTransport/GDTCORLibrary/Public/GDTCORStorageProtocol.h:41]: That's not a bad idea--but these are methods that I want to require of all future storage implementations.

File: GoogleDataTransport/GDTCORLibrary/Public/GDTCORStorageProtocol.h (hunk 5 above)
Comment (mikehaney24) [reply GoogleDataTransport/GDTCORLibrary/Public/GDTCORStorageProtocol.h:47]: It's meant to store library state. A caller _can_ implement NSSecureCoding to generate its bytes, but it's not a hard requirement. How the caller generates bytes is a decision left to that specific class. Because several of the callers will be singletons, it's created a lot of NSCoding weirdness that this can help to avoid. I prefer to keep it as NSData because then NSCoding can be used when wanted, or, it can just be a string or whatever.

File: GoogleDataTransport/GDTCORLibrary/Public/GDTCORStorageProtocol.h (hunk 3 above)
Comment (mikehaney24) [reply GoogleDataTransport/GDTCORLibrary/Public/GDTCORStorageProtocol.h]: Done.

File: GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m (hunk 2 above)
Comment (mikehaney24) [reply GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m:54]: You're right, but the way it is is more consistent with other similar folder functions in GDT.

**Comment by paulb777** [comment]
Please fix GHA before merging.
